File: python/trees/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        answers = []

        def dfs(node):
            if not node:
                return 0

            left = dfs(node.left)
            right = dfs(node.right)
            answers.append(max(left + node.val, right + node.val, node.val, left + right + node.val))

            return max(left + node.val, right + node.val, node.val)

        dfs(root)

        return max(answers)

File: c++/trees/binary-tree-maximum-path-sum.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        return findSum(root).first;
    }

private:
    std::pair<int, int> findSum(TreeNode* node) {
        if (!node) {
            return std::make_pair(std::numeric_limits<int>::min(), std::numeric_limits<int>::min());
        }

        std::pair<int, int> left = findSum(node->left);
        std::pair<int, int> right = findSum(node->right);

        return std::make_pair(std::max({left.first, right.first, std::max(left.second, 0) + std::max(right.second, 0) + node->val}), std::max({std::max(left.second, 0) + node->val, std::max(right.second, 0) + node->val}));
    }
};
